### ISSUE TITLE
What do I do when I'm done with the editor?

### ISSUE THREAD

**Issue by benlong100** [stevengharris/MarkupEditor#165]
I don't keep the MarkupEditor visible and active all the time in my app. When I'm done with it, is there anything I need to do to tidy up? I'm having some weird issues and am wondering if some of the various coordinators might be up to something.

**Comment by stevengharris** [comment]
No, it should not be leaving any cruft around. I am using MarkupEditorViews in a List and have not noticed any issues. Please let me know if you discover something that needs cleaning up.

**Comment by benlong100** [comment]
Okay, I expect this is just me doing things wrong. I have a save button that sits outside of MarkupEditor. I need to get modified HTML out of the editor so that I can do a bunch of stuff to it before saving it so my button is saying:

` MarkupEditor.selectedWebView?.getHtml {html in
            theWYSIWYGHTML = html ?? ""
         }`

I'm not consistently getting anything into theWYSIWYGHTML. Possibly because the webView is not selected at the time? Is there a better way for me to extract the edited stuff?

**Comment by stevengharris** [comment]
I use an "autosave" type of approach. In `MarkupDelegate.markupDidLoad`, I set up a callback that I invoke at time intervals when the contents has changed. The contents for me is held by a model object that also tracks its "save state". The callback takes the `view` passed in `MarkupDelegate.markupDidLoad` as an argument, so it doesn't need to depend on the optional `selectedWebView`. Then in `MarkupDelegate.markupInput`, I flag the model object as being modified, and I reset the timer on autosave. The autosaver then always saves at some maximum interval if the model object has been modified, but it is not consuming resources when no input is being received. My autosave interval is short enough that I haven't done a kind of force-save when, for example, the app is closing, but I've considered doing that also.

Roughly, the autosave looks like (I have some other things going on with sync to iCloud that I deal with separately from the autosave):

```
    /// The timer that delayed saving triggered.  Save the contents of the MarkupWebView.
    private func autosave(from view: MarkupWKWebView) {
        view.getHtml { contents in
            guard let contents else {
               <myModelObject>.savedState = .error
                return
            }
            <myModelObject>.contents = contents
            do {
                try <myModelObject>.save()
            } catch let error {
                Logger.model.error("Error autosaving: \(error.localizedDescription)")
                <myModelObject>.savedState = .error
            }
        }
    }
```

**Comment by benlong100** [comment]
Thanks, that was very helpful. I really like the idea of the timed save, and will implement that soon. In the meantime, I tried connecting my button to the view that I captured from markupDidLoad, but it had the same problems: it would only work intermittently. Then I noticed that, actually, it would work once and then stop working and not start again until I'd cleaned the project and deleted all my breakpoints. I tried it on another computer and it was fine, so something was just screwy with my first machine. As such, it might be that my original scheme of referencing MarkupEditor.selectedWebView would have worked. But this seems sturdier!
